roachprod-microbench: confidence interval calculation

herkolategan opened the pull request:
This PR replaces the existing `benchseries` implementation with a new metrics model for populating the microbenchmark dashboards. The previous approach was difficult to maintain, prone to panics due to missing metrics, and required excessive boilerplate code to format the metrics for `benchseries`.

The updated implementation now uses an internal confidence interval calculation integrated within the metrics builder, eliminating the need for additional computation steps through `benchseries`.

Additionally, this change introduces the bootstrap confidence interval calculation for two sets of microbenchmark results. This method allows for better estimation of performance uncertainty by resampling benchmark results. It provides a distribution of outcomes that accounts for performance variability and system fluctuations without assuming normality.

Epic: None
Release note: None

herkolategan committed:
roachprod-microbench: confidence interval calculation

This change adds the ability to calculate the bootstrap confidence interval given two sets
of microbenchmark results.

When applying the bootstrap confidence interval method to microbenchmarks, it
helps estimate the uncertainty in performance measurements. By resampling
benchmark results with replacement and computing the statistic (e.g., mean) over
multiple iterations, we obtain a distribution of possible outcomes. The
confidence interval derived from this distribution provides a robust estimate of
performance variability, accounting for noise and system fluctuations without
assuming normality.

Epic: None
Release note: None

herkolategan committed:
roachprod-microbench: replace benchseries with metrics model

Previously, to populate the microbenchmark dashboards the `benchseries`
implementation was utilised to calculate the confidence interval to plot for the
microbenchmarks dashboard. This implementation became hard to maintain as it
often panicked if a metric was missing from one set, and required boilerplate
code to feed the metrics in specific ways to adhere to the form `benchseries`
expects it.

This change updates the `push to influx` implementation to use our own
confidence interval calculation, that has been built into the metrics builder,
and now does not require an additional calculation step.

Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/141990)
<!-- Reviewable:end -->

srosenberg commented on pkg/cmd/roachprod-microbench/model/math.go:
```diff
@@ -0,0 +1,98 @@
+// of two sets of values. The confidence interval is calculated using a
+// bootstrap method.
+func calculateConfidenceInterval(newValues, oldValues []float64) ConfidenceInterval {
+	rng := rand.New(rand.NewSource(hash(newValues) + hash(oldValues)))
```
Why `hash` instead of grabbing the wall clock or `/dev/urandom`? Hashes commute, so swapping the args, would yield the same seed... don't we want independent resampling of the same input?

herkolategan replied:
I don't think we ever re-run a comparison to see if it yields different results. This is a one-off calculation per comparison. We can pass in an RNG here, which would be required to make the unit tests pass deterministically, but I don't see the benefit of having varying results when we could have deterministic comparisons yielding the same result for the same dataset?

> so swapping the args, would yield the same seed

True, but once again I don't see an issue with this since we just want a seed for the bootstrap calc, and having a deterministic seed here is not problematic, I think? If the sample sets are swapped the delta is just the inverse as well.

srosenberg replied:
Makes sense. I don't think determinism causes an issue here.

srosenberg commented on pkg/cmd/roachprod-microbench/model/math.go:
```diff
@@ -0,0 +1,98 @@
+// calculateConfidenceInterval calculates the confidence interval for the ratio
+// of two sets of values. The confidence interval is calculated using a
+// bootstrap method.
+func calculateConfidenceInterval(newValues, oldValues []float64) ConfidenceInterval {
```
We probably want to bail out if either input is empty, otherwise median will be out-of-bounds.

herkolategan replied:
Makes sense, I'll add a check here.

srosenberg commented on pkg/cmd/roachprod-microbench/model/math.go:
```diff
@@ -0,0 +1,98 @@
+		resample(rng, newValues, resNew)
+		resample(rng, oldValues, resOld)
+		medOld := median(resOld)
+		if medOld == 0 {
```
Not super familiar with the bootstrap method... wouldn't this introduce bias if this happens _frequently_?

srosenberg commented on pkg/cmd/roachprod-microbench/model/math.go:
```diff
@@ -0,0 +1,98 @@
+const resampleCount = 1000
+const confidence = 0.95
+
+// calculateConfidenceInterval calculates the confidence interval for the ratio
```
Ideally, we wouldn't have to roll our own. At minimum, we should cover this with some sort of a randomized test suite, preferably with a trusted oracle.

herkolategan replied:
As far as testing goes, I eye-balled it against a set that ran through `benchseries`, but I could write a unit test that tests our own against `benchseries` to confirm the results are correct. I'll add one.

I would have rather preferred if `benchseries` was more accessible to use the parts we need. Formatting the data to fit the ingestion point became an issue when we added and removed a metric of a benchmark `err/ops` which causes a panic in `benchseries`.  Not sure how Google handles this since I'm relatively sure it would have panicked there as well, unless I'm missing some way to prevent the panic, or they do some kind of preprocessing.

herkolategan replied:
In regards to rolling our own (I agree), but the calculation deviates enough that implementing a small version of it versus pulling in a stats library might be more worthwhile? I also couldn't really find something publicly available that uses bootstrap and the median for comparing two sets.

srosenberg replied:
Makes sense; I also couldn't find an existing implementation in Go.

herkolategan committed:
roachprod-microbench: add confidence interval oracle test

This test compares the result of the benchseries method with the newly added
internal method. Since there is randomization in the bootstrap, a small
tolerance is used to compare the results.

srosenberg commented on pkg/cmd/roachprod-microbench/model/math.go:
```diff
@@ -0,0 +1,98 @@
+	for i := range dest {
+		dest[i] = src[r.Intn(l)]
+	}
+	sort.Float64s(dest)
```
Technically, resampling doesn't require sorting. It's an implementation detail of `median`, so for the sake of separation of concerns, we should move it there. (Incidentally, sorting isn't strictly needed for `median` either; a recursive quick-select is O(n), but it's probably an overkill since we don't expect large input sizes.)

herkolategan replied:
Good point. Going for the slow readable solution, since the input size is small.

srosenberg commented on pkg/cmd/roachprod-microbench/model/math_test.go:
```diff
@@ -0,0 +1,130 @@
+			// Assert that the two methods give the same result, within a small tolerance.
+			// The tolerance is due to the randomization in the bootstrap.
+			const tolerance = 0.01
+			require.InDelta(t, oldCI.Low, newCI.Low, tolerance, "confidence interval lows differ")
```
Did you `--stress` this test? I seem to recall a failure when I was playing around with the PR, but it could have been due to local changes.

herkolategan replied:
I did, but will run a 10k count before merging to make sure.

herkolategan commented:
TFTR!

bors r=srosenberg
